Full treewide Nix format and enforcement [skip treewide]

infinisil opened the pull request:
> [!Note]
> A merge party for final review and merge will be held on 2025-04-01 ~19:00~ 20:00 CEST in [this Jitsi](https://meet.zrh.init7.net/nix-formatting), anybody is free to join! :tada:
> 
> See https://discourse.nixos.org/t/nix-formatting-team-full-nixpkgs-reformat/61867 for more info

This is a huge step towards the full implementation of NixOS/rfcs#166 (ref https://github.com/NixOS/nixfmt/issues/153)! Highlights:
- You can now regularly run `treefmt` inside `nix-shell`/`nix develop` or `nix fmt` to get the right formatting on all files, no more "these particular files need to be formatted"
- You can even hook up your editor's to run `nixfmt` on auto-save or set up a pre commit hook to run `treefmt`
- CI will enforce this on all PRs going forward, ensuring that from now on, all Nix files consistently use the same formatting

This PR _will_ lead to merge conflicts for a number of PRs, up to an estimated ~1100 (~33%) among the PRs with activity in the past 2 months, but that should be lower than what it would be without the previous partial reformat (https://github.com/NixOS/nixpkgs/pull/322537).

Merge conflicts caused by this commit can now also automatically be resolved while rebasing using the [auto-rebase script](https://github.com/NixOS/nixpkgs/tree/8616af08d915377bd930395f3b700a0e93d08728/maintainers/scripts/auto-rebase).

This PR should be reviewed and approved by the @NixOS/nix-formatting team.
## Things done
- [x] Tested `treefmt` in `nix-shell`/`nix develop`
- [x] Tested `nix fmt`
- [x] Update/add docs
- [x] Do the treewide reformat (but needs constant updating as master gets updated)
- [x] Added the reformat commit to `.git-blame-ignore-revs` (also needs constant updating)
  - [x] Check that the auto-rebase script works
  - [x] Check that GitHub ignores the commit (it's highlighted in bold red, a bad sign.. Edit: All good: https://github.com/Nixinova/NovaGrammars/pull/11)
- [x] Tested the GHA workflow to check Nix formatting: https://github.com/tweag/nixpkgs/pull/109
- [x] Investigated the causes for the [changed paths](https://github.com/NixOS/nixpkgs/actions/runs/13293497183?pr=380990#summary-37120020657)
  - Nothing problematic, but I opened https://github.com/NixOS/nixpkgs/pull/382880 for a small orthogonal improvement
- [x] Confirm it uses the latest treefmt version (relates to https://github.com/NixOS/nixpkgs/pull/387745)
- [x] Make sure the entire tree gets formatted, even if the commands are run in subdirectories
- [x] Prepare tooling to help with automated conflict resolution
  - [x] Implement https://github.com/NixOS/nixfmt/pull/277 -> https://github.com/NixOS/nixfmt/pull/283
  - [x] Update `pkgs.nixfmt-rfc-style` to include the above: https://github.com/NixOS/nixpkgs/pull/387047
  - [x] Update CI's nixfmt version: https://github.com/NixOS/nixpkgs/pull/389635
- [x] Create an issue to describe how to resolve merge conflicts, pin it: https://github.com/NixOS/nixpkgs/issues/395247
- [x] Announce it on Discourse: https://discourse.nixos.org/t/nix-formatting-team-full-nixpkgs-reformat/61867
- [x] Prepare for 24.11 backport: https://github.com/NixOS/nixpkgs/pull/395018
- [x] Make sure staging isn't a problem: Merging will give conflicts, so needs to be done manually, but should be easy with the [mergetool](https://github.com/NixOS/nixfmt?tab=readme-ov-file#git-mergetool) followed by a `treefmt` as an extra check.

Potential follow-up: There are cases where [new files could be committed with the wrong formatting](https://github.com/NixOS/nixpkgs/issues/322520#issuecomment-2576284686). If this becomes a sufficiently annoying problem, we should have some automation to automatically fix this. But let's see about that when we get there.

---

This work is funded by [Tweag](https://tweag.io) and [Antithesis](https://antithesis.com/) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
shell: Introduce treefmt

Introduces treefmt with a simple nixfmt-rfc-style configuration to
format all Nix files.

This is only practically usable with the following commit that formats
all files accordingly.

infinisil committed:
flake.nix: Set formatter

This enables `nix fmt`, though it won't be practically usable without
also reformatting all files, which is done in a following commit.

infinisil committed:
workflows/check-nix-format: Enforce formatting on all files

Changes the Nix format checking workflow to now strictly enforce
formatting of all Nix files using the treefmt setup introduced
in the pre-previous commit.

This is in [accordance with the approved RFC 166](https://github.com/NixOS/rfcs/blob/master/rfcs/0166-nix-formatting.md#reformat-nixpkgs).

Note that the "skip treewide" thing is no longer necessary, already
before, because there's nothing that would fail for treewide changes.
Previously the problem was that the GitHub API would be bombarded.

infinisil committed:
CONTRIBUTING: Improve and update formatting docs

SuperSandro2000 commented:
> Merge conflicts caused by this commit can now also automatically be resolved while rebasing using the [auto-rebase script](https://github.com/NixOS/nixpkgs/tree/8616af08d915377bd930395f3b700a0e93d08728/maintainers/scripts/auto-rebase).

Can you point me to the instructions for that? I can't seem to find them.

infinisil commented on shell.nix:
```diff
@@ -28,10 +28,10 @@ let
   packages = with pkgs; [
-    # The default formatter for Nix code
-    # See https://github.com/NixOS/nixfmt
-    nixfmt-rfc-style
```
They can, the treefmt wrapper includes nixfmt :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-formatting-team-full-nixpkgs-reformat/61867/1

infinisil commented:
@SuperSandro2000 It's really just doing `./run.sh` as mentioned in the docs there. We now also have it described in the new Discourse announcement though.

---

We have now scheduled this PR for final review and merging in the meeting on Tuesday 2025-04-01 19:00 CEST ([Jitsi](https://meet.zrh.init7.net/nix-formatting)), anybody is free to join!

Note that I'll need to rebase it and merge it right after to avoid merge conflicts. If anybody sees any issues with this PR, let me know in advance :)

infinisil commented:
We’re on schedule to merge this tomorrow! I messed up the times a little bit all over the place, so just to clarify: We meet
Tomorrow 8:00 PM in [this Jitsi](https://meet.zrh.init7.net/nix-formatting) for the merge party :tada:

MattSturgeon commented on ci/default.nix:
```diff
@@ -21,9 +21,42 @@ let
+
+  fmt =
+    let
+      treefmtNixSrc = fetchTarball {
+        # Master at 2025-02-12
+        url = "https://github.com/numtide/treefmt-nix/archive/4f09b473c936d41582dd744e19f34ec27592c5fd.tar.gz";
+        sha256 = "051vh6raskrxw5k6jncm8zbk9fhbzgm1gxpq9gm5xw1b6wgbgcna";
+      };
```
Now we have `treefmt.withConfig` and/or `nixfmt-tree` I don't think we need this dependency on treefmt-nix.

Or maybe `config.build.devShell` is doing something important, that adding `treefmt.withConfig {}` to a shell wouldn't do?

(This is minor enough to delay for another PR, especially given how late my comment is)

MattSturgeon commented on CONTRIBUTING.md:
````diff
@@ -531,14 +531,31 @@ If you removed packages or made some major NixOS changes, write about it in the
+```
+
+If you're starting your editor in `nix-shell` or `nix develop`,
+you can also set it up to automatically format the file with `nixfmt` or `treefmt` on save.
````
Q: do we want to encourage running `nixfmt` manually or should we only document `treefmt`?

infinisil committed:
treewide: Format all Nix files

Format all Nix files using the officially approved formatter,
making the CI check introduced in the previous commit succeed:

  nix-build ci -A fmt.check

This is the next step of the of the [implementation](https://github.com/NixOS/nixfmt/issues/153)
of the accepted [RFC 166](https://github.com/NixOS/rfcs/pull/166).

This commit will lead to merge conflicts for a number of PRs,
up to an estimated ~1100 (~33%) among the PRs with activity in the past 2
months, but that should be lower than what it would be without the previous
[partial treewide format](https://github.com/NixOS/nixpkgs/pull/322537).

Merge conflicts caused by this commit can now automatically be resolved while rebasing using the
[auto-rebase script](https://github.com/NixOS/nixpkgs/tree/8616af08d915377bd930395f3b700a0e93d08728/maintainers/scripts/auto-rebase).

If you run into any problems regarding any of this, please reach out to the
[formatting team](https://nixos.org/community/teams/formatting/) by
pinging @NixOS/nix-formatting.

infinisil committed:
git-blame-ignore-revs: Add previous commit

And enable use of the auto-rebase script: https://github.com/NixOS/nixpkgs/tree/master/maintainers/scripts/auto-rebase

infinisil commented:
Reviewed together in a meeting with ~10 people, determined that there are no blockers, going to merge!

jfly reviewed (approved):
:shipit:

mightyiam commented:
:balloon:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/formatting-team-meeting-2025-04-01/62524/1

infinisil commented:
Backport was done manually here: https://github.com/NixOS/nixpkgs/pull/395018
